Add mantidworkbench as an entry point for the mantidworkbench Conda package

thomashampson opened the pull request:
### Description of work
- Add `mantidworkbench` as an entry point to the application. Users can now launch with either `workbench` or `mantidworkbench` from within a Conda environment.
- Configure the Conda package entry points to use `jemalloc` on Linux. Previously, the `workbench` entry point did not include the same `jemalloc` setup as the `mantidworkbench` script that was available on Linux installs. Now on Linux both entry points will call the same launch script that will configure `jemalloc` to be used.
Here is a diagram to help illustrate the changes:
![image](https://github.com/user-attachments/assets/b4a3df23-a557-4194-a9df-70622ace6668)


#### Summary of work
Added a new wrapper script called `workbench_launch_wrapper` that is now the entry point defined in the conda-build recipe. The new script will detect whether you are using Linux. If not, workbench is launched as before. If it is Linux, it will call the script that used to be named `mantidworkbench` that will configure `jemalloc` before launching the application.

#### Purpose of work
- There have been cases where users have asked for a `mantidworkbench` entry point to avoid confusion.
- It is already an entry point on Linux, but will perform additional steps before launching the application (e.g. setting up `jemalloc`)
- Currently if you launch on Linux with the Conda entry point (`workbench`), it bypasses the `jemalloc` setup, and the application does not perform well.

Fixes #37538 

### To test:
On **all OS**:
Install the test package:
`conda install jclarkestfc/label/entryPointTest::mantidworkbench`
Test the following things:
1. Verify that mantid workbench launches with both `mantidworkbench` and `workbench`
2. Test that the error report functions with both entry points (e.g. use the SegFault algorithm)
3. Test that the error reporter doesn't pop up when passing the `--single-process` argument to either entry point
4. (Linux only) verify that both launch entry points configure jemalloc correctly by running `os.environ['LD_PRELOAD']` in workbench's python console. It should output the location of `libjemalloc.so.2` from inside your Conda environment.
5. (Linux only) Check that the --debug argument works when launching with either entry point.
6. On each OS, install the standalone packages and make sure they still work as expected:
https://builds.mantidproject.org/job/build_packages_from_branch/957/

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jhaigh0 committed:
Add new wrapper script to unify launch commands

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Thomas Hampson <thomas.hampson@stfc.ac.uk>

jhaigh0 committed:
Add arg passing to workbench launch wrapper script

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Thomas Hampson <thomas.hampson@stfc.ac.uk>

jhaigh0 committed:
Add release notes

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Thomas Hampson <thomas.hampson@stfc.ac.uk>

thomashampson committed:
Update launch script name in linux standalone packaging script

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Change python exec command to avoid infinite loop when launching on Linux

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

thomashampson committed:
Fix name of mantid launch script in standalone package script

Co-authored-by: James Clarke <james.clarke@stfc.ac.uk>
Co-authored-by: Jonathan Haigh <jonathan.haigh@stfc.ac.uk>

peterfpeterson commented on qt/applications/workbench/workbench/app/workbench_launch_wrapper.py:
```diff
@@ -0,0 +1,23 @@
+from workbench.app.main import main
+
+
+def launch():
```
I think this is missing the bits for return codes. The final lines should be like [this](https://github.com/neutrons/SNAPRed/blob/next/src/snapred/__main__.py#L139-L140).

peterfpeterson commented on buildconfig/CMake/LinuxPackageScripts.cmake:
```diff
@@ -100,10 +100,11 @@ set(PYTHON_ARGS " -Wdefault::DeprecationWarning -Werror:::mantid -Werror:::manti
 set(PREAMBLE "${CONDA_PREAMBLE_TEXT}")
 set(LOCAL_PYPATH "${CMAKE_CURRENT_BINARY_DIR}/bin/")
+# The python command to start workbench
+set(MANTIDWORKBENCH_EXEC "-c \"from workbench.app.main import main; main()\"")
```
We appear to be most of the way there https://github.com/mantidproject/mantid/blob/main/qt/applications/workbench/workbench/__main__.py

peterfpeterson commented:
Can the new wrapper get named `mantidworkbench_whatever` instead? `workbench` in the `PATH` is pretty ambigious.

sf1919 commented:
I'm going to change the milestone on this as it's unlikely we can get this sorted before end of tomorrow.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jclarkeSTFC committed:
Move release notes to 6.12 folder

jclarkeSTFC committed:
Add return code handling

jclarkeSTFC committed:
Rename workbench_launch_wrapper to mantidworkbench_launch_wrapper

jclarkeSTFC committed:
Rename standalone launcher script

Renamed from workbench_launcher_jemalloc to launch_mantidworkbench

peterfpeterson commented on qt/applications/workbench/workbench/app/mantidworkbench_launch_wrapper.py:
```diff
@@ -0,0 +1,23 @@
+from workbench.app.main import main
+
+
+def launch():
```
This becomes much easier to add tests to if the `sys.argv` is an argument to the function
```suggestion
def launch(args):
```

peterfpeterson commented on qt/applications/workbench/workbench/app/mantidworkbench_launch_wrapper.py:
```diff
@@ -0,0 +1,23 @@
+
+
+if __name__ == "__main__":
+    sys.exit(launch())
```
then this becomes
```suggestion
    sys.exit(launch(sys.argv))
```

peterfpeterson commented on qt/applications/workbench/workbench/app/mantidworkbench_launch_wrapper.py:
```diff
@@ -0,0 +1,23 @@
+        command = ["launch_mantidworkbench"] + sys.argv[1:]
+        subprocess.run(command)
+    else:
+        main()
```
shouldn't `main()` get the command line arguments?

jclarkeSTFC committed:
Pass command-line arguments through to main

Co-authored-by: Pete Peterson <petersonpf@ornl.gov>

thomashampson commented:
I've updated the diagram in the description to reflect the recent changes.

thomashampson commented:
I've verified on Windows that
- both entry points work from a Conda install
- the standalone install launches fine

thomashampson commented:
I've verified that `os.environ['LD_PRELOAD']` outputs the expected jemalloc location on Linux conda {workbench, mantidworkbench} and standalone package.

thomashampson commented:
For some reason the standalone package isn't launching on MacOS.

jclarkeSTFC commented:
https://builds.mantidproject.org/job/build_packages_from_branch/956/

macOS standalone after rebase

jhaigh0 reviewed (commented):
Whoops didn't mean to approve yet

jhaigh0 reviewed (changes requested):
(fix the mac stand alone)

jclarkeSTFC committed:
Add optional argument to workbench main() and launch()

Then someone can pass in arguments directly instead of using
sys.argv. If no arguments are given then sys.argv gets used
by default.

Co-authored-by: Thomas Hampson <thomas.hampson@stfc.ac.uk>

jclarkeSTFC commented:
https://builds.mantidproject.org/job/build_packages_from_branch/957/

This one's going to be a winner

peterfpeterson commented on qt/applications/workbench/workbench/app/main.py:
```diff
@@ -12,7 +12,7 @@
 
 
-def main():
+def main(args=None):
```
I like that you modified this to make it easier to test automatically

peterfpeterson reviewed (dismissed):
The changes are reasonable and I have tried out various thing for a developer build in linux which work as expected.

sf1919 commented:
@jhaigh0 Can you look at this please? It's saying you requested changes so can't be gatekeepered.

cailafinn commented on docs/source/release/v6.12.0/Workbench/New_features/37538.rst:
```diff
@@ -0,0 +1,2 @@
+- Add `mantidworkbench` as an entry point to launch the mantidworkbench Conda package.
+- Configure the Conda `workbench` and `mantidworkbench` entry points to launch workbench with jemalloc on Linux.
```
RST needs double backticks for code formatting:
```suggestion
- Add ``mantidworkbench`` as an entry point to launch the mantidworkbench Conda package.
- Configure the Conda `workbench` and ``mantidworkbench`` entry points to launch workbench with jemalloc on Linux.
```

thomashampson replied:
I edited your suggestion to add double ticks around the workbench as well.

cailafinn reviewed (changes requested):
Changes look good, just want to check we actively decided not to do make this change to the exec rather than it just being missed for this PR: 
- https://github.com/mantidproject/mantid/pull/37750/files#r1709824134

thomashampson committed:
add backticks to release notes

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>
